src/orchestrator-kubernetes/src/cloud_resource_controller.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Management of K8S objects, such as VpcEndpoints.
+
+use std::collections::HashSet;
+use std::str::FromStr;
+
+use async_trait::async_trait;
+use kube::api::{DeleteParams, ListParams, ObjectMeta, Patch, PatchParams};
+use kube::ResourceExt;
+
+use maplit::btreemap;
+use mz_cloud_resources::crd::vpc_endpoint::v1::{VpcEndpoint, VpcEndpointSpec};
+use mz_cloud_resources::CloudResourceController;
+use mz_repr::GlobalId;
+
+use crate::{KubernetesOrchestrator, FIELD_MANAGER};
+
+fn vpc_endpoint_name(id: GlobalId) -> String {
+    format!("connection-{id}")
+}
+
+#[async_trait]
+impl CloudResourceController for KubernetesOrchestrator {
+    async fn ensure_vpc_endpoint(
+        &self,
+        id: GlobalId,
+        spec: VpcEndpointSpec,
+    ) -> Result<(), anyhow::Error> {
+        let name = vpc_endpoint_name(id);
+        let mut labels = btreemap! {
+            "environmentd.materialize.cloud/connection-id".to_owned() => id.to_string(),
+        };
+        for (key, value) in &self.config.service_labels {
+            labels.insert(key.clone(), value.clone());
+        }
+        let vpc_endpoint = VpcEndpoint {
+            metadata: ObjectMeta {
+                labels: Some(labels),
+                name: Some(name.clone()),
+                namespace: Some(self.kubernetes_namespace.clone()),
+                // TODO owner references?
+                //owner_references: todo!(),
```

Comment:
Did you mean to leave this in? I'm not really sure what it means.

Reply:
I've created https://github.com/MaterializeInc/cloud/issues/4408 to track that.

Owner references are K8S's way of tying objects with their parents. I think we should probably add these to everything created by environmentd, but it's not critical, since they _should_ be cleaned up when the namespace is deleted.